### DISCUSSION TITLE
Feature Request: Read/Import Fixed Width Files

### DISCUSSION THREAD

**Discussion by szimmer** [duckdb/duckdb#13777] [Ideas]
Some flat files are fixed width with columns determined by their position in the file. It would be nice to parse these by specifying column names and the widths.

**Comment by rplsmn** [comment]
Seconding this request !
Although with some metaprogramming and substr we get by, would love to see something akin to read_fwf from Python's pandas or  R's 
readr

**Comment by lakemove** [comment]
maybe could use #16975 

```
from regexp_read_table(file_content, '(?P<col1>.{3})(?P<col2>.{4}')
```

**Comment by nicosuave** [comment]
This sounds like a fun extension project...

**Comment by xevix** [comment]
I might have a look at this as I've wanted this as well for dimensions data that comes in fixed-width CSVs #16008

**Reply by xevix** [reply]
@pdet By any chance have you done any work on this as part of the CSV Sniffer? I was thinking of trying to tackle this, since it would be very useful to have. I think it's not RFC 4180 compliant but I think it's common enough to be valuable.

**Reply by pdet** [reply]
The sniffer does not detect column widths. At one point, I considered calculating the average row width for cardinality estimation purposes, but I don't believe I ever implemented it. That said, I think it would be straightforward to extend the dialect detection to also check whether columns have consistent widths and determine their specific widths.

**Reply by xevix** [reply]
Ah ok, I wasn't sure how difficult this would be as I've only glanced at the sniffer code. I was thinking a few things:

1. Detect fixed-width dialect, either autodetect or via an option
2. If header is enabled, and this is fixed-width, use the header to determine name and starting positions for the columns
3. If schema is passed in which includes column names and starting position, use that to read the file

Does the above sound reasonable? Would you be interested in taking this on? If not, I'd be interested in giving it a shot, if the idea sounds good.

**Reply by pdet** [reply]
I don't think Ill be able to tackle this in the foreseeable future, so feel free to have a go, im happy to review a PR :-)

Besides the points you raised, what is really interesting about fixed width columns is that you can be much more efficient with parallelization, projection pushdown, and more accurate on file cardinality. So I believe these would also need to be taken into consideration. And they might be slightly more complex to implement than the sniffer changes.

**Reply by xevix** [reply]
Sounds good, I'll give it a shot! This may take a bit to get familiarized with the CSV code.

Yeah, I think maybe there's a chance for some efficiency gains just for this format. Since the structure of each row is theoretically identical, once the structure is known processing can be faster.

I was thinking about the algorithm for schema detection, and it seemed as though for the general case, you have to read the entire file to get the format, which would nullify some of these gains.

1. Initialize a vector<boolean> to track if there's whitespace in all the lines of a file at the given character position, default to true
2. For each line, if a non-space character is detected, set the boolean at that position to false
3. Once file is processed, check the vector<boolean> for runs of at least 1 non-space character surrounded by 1 space character on left and right side (or start or end of line), and if so, consider this to be a column

Fixed width refers to number of characters and not bytes, so I'd have to think about opportunities for breaking up the file for processing in a way that works. I think you'd still have to check for newlines in order to batch into some number of lines to break up the work.

Notes/Concerns
* Columns with no values, if multiple of them are back to back, cannot be discerned from one another. Maybe this is ok though, since if they're entirely blank, probably won't do much analysis on them. Also, the resulting comma positions could be output so a user could then pass the structure next time around, or even `COPY` to a separator-based CSV
* Telling between separator-based and width-based columns is tricky, need to think of good heuristics here, or have width-based be option-enabled to be unambiguous
* It would be nice to have column names auto-generated if auto-detecting schema and header=false, but this might conflict with how existing CSV parsing works

**Reply by szimmer** [reply]
Just a note - most fixed width files I've ever worked with have very little whitespace. They are saved this way (historically) to not use up memory on things like delimiters and space. And if you know the schema, they can be read in very quick when you want to only specify a few columns. There's no parsing, in a way, just a schema telling you which variable is in which column.

**Reply by xevix** [reply]
That makes a lot of sense why they would use that format then. Yeah I was mostly talking about the autodetection, but perhaps a good starting point for this work would be supporting fixed-width in general with a mandatory schema passed in to start, which would at least enable reading the files.

**Reply by xevix** [reply]
For some examples of sparse fixed-width files, here's some examples here (extension `.txt`) https://noaa-ghcn-pds.s3.amazonaws.com/index.html

```
In [6]: df = pd.read_fwf('ghcnd-stations.txt', header=None)

In [7]: duckdb.sql("SUMMARIZE FROM df").show()
┌─────────────┬─────────────┬────────────────────┬─────────────┬───────────────┬────────────────────┬────────────────────┬────────────────────┬─────────────────────┬────────────────────┬────────┬─────────────────┐
│ column_name │ column_type │        min         │     max     │ approx_unique │        avg         │        std         │        q25         │         q50         │        q75         │ count  │ null_percentage │
│   varchar   │   varchar   │      varchar       │   varchar   │     int64     │      varchar       │      varchar       │      varchar       │       varchar       │      varchar       │ int64  │  decimal(9,2)   │
├─────────────┼─────────────┼────────────────────┼─────────────┼───────────────┼────────────────────┼────────────────────┼────────────────────┼─────────────────────┼────────────────────┼────────┼─────────────────┤
│ 0           │ VARCHAR     │ ACW00011604        │ ZI000067991 │        111125 │ NULL               │ NULL               │ NULL               │ NULL                │ NULL               │ 129657 │            0.00 │
│ 1           │ DOUBLE      │ -9.99              │ 90.0        │         83239 │ 36.46672134246484  │ 11.741397897114103 │ 30.766161947246953 │ 37.63219878764216   │ 42.92032386979219  │ 129657 │            0.00 │
│ 2           │ DOUBLE      │ -99.9996           │ 179.9947    │         85264 │ 12.463500342442547 │ 98.81871214160806  │ -84.7080026814677  │ -11.426538325908751 │ 112.53416320638549 │ 129657 │            0.00 │
│ 3           │ DOUBLE      │ -999.9             │ 5033.0      │         10474 │ 445.57069807261456 │ 651.7748498800127  │ 86.59087078290393  │ 257.62649877485995  │ 588.192127568991   │ 129657 │            0.00 │
│ 4           │ VARCHAR     │ "HOME" MONOGORILBY │ Zirkel      │        101592 │ NULL               │ NULL               │ NULL               │ NULL                │ NULL               │ 129657 │            0.00 │
│ 5           │ VARCHAR     │ &R.                │ ZAC         │           978 │ NULL               │ NULL               │ NULL               │ NULL                │ NULL               │ 129657 │           98.35 │
│ 6           │ VARCHAR     │ GSN                │ GSN         │             1 │ NULL               │ NULL               │ NULL               │ NULL                │ NULL               │ 129657 │           99.24 │
│ 7           │ DOUBLE      │ 1001.0             │ 99999.0     │          6687 │ 55235.1683404469   │ 26569.622608861097 │ 33935.722990246235 │ 58160.921869566904  │ 72584.6584736155   │ 129657 │           93.86 │
└─────────────┴─────────────┴────────────────────┴─────────────┴───────────────┴────────────────────┴────────────────────┴────────────────────┴─────────────────────┴────────────────────┴────────┴─────────────────┘


In [8]: df = pd.read_fwf('ghcnd-stations.txt', header=None, infer_nrows=200_000)

In [9]: duckdb.sql("SUMMARIZE FROM df").show()
┌─────────────┬─────────────┬────────────────────┬─────────────┬───────────────┬────────────────────┬────────────────────┬─────────────────────┬────────────────────┬─────────────────────┬────────┬─────────────────┐
│ column_name │ column_type │        min         │     max     │ approx_unique │        avg         │        std         │         q25         │        q50         │         q75         │ count  │ null_percentage │
│   varchar   │   varchar   │      varchar       │   varchar   │     int64     │      varchar       │      varchar       │       varchar       │      varchar       │       varchar       │ int64  │  decimal(9,2)   │
├─────────────┼─────────────┼────────────────────┼─────────────┼───────────────┼────────────────────┼────────────────────┼─────────────────────┼────────────────────┼─────────────────────┼────────┼─────────────────┤
│ 0           │ VARCHAR     │ ACW00011604        │ ZI000067991 │        111125 │ NULL               │ NULL               │ NULL                │ NULL               │ NULL                │ 129657 │            0.00 │
│ 1           │ DOUBLE      │ -90.0              │ 83.65       │         81556 │ 26.120366526296436 │ 28.02523214774972  │ 22.02789194579678   │ 36.99985321783737  │ 42.79220922606566   │ 129657 │            0.00 │
│ 2           │ DOUBLE      │ -179.9947          │ 179.7414    │         82324 │ -45.67528928866157 │ 88.51118441234814  │ -100.35778055568557 │ -85.19082441544143 │ -25.904367956196637 │ 129657 │            0.00 │
│ 3           │ DOUBLE      │ -999.9             │ 5033.0      │         10474 │ 445.57069807261456 │ 651.7748498800127  │ 86.59087078290393   │ 257.62649877485995 │ 588.192127568991    │ 129657 │            0.00 │
│ 4           │ VARCHAR     │ AB                 │ YT          │            82 │ NULL               │ NULL               │ NULL                │ NULL               │ NULL                │ 129657 │           33.93 │
│ 5           │ VARCHAR     │ "HOME" MONOGORILBY │ Zirkel      │        111157 │ NULL               │ NULL               │ NULL                │ NULL               │ NULL                │ 129657 │            0.00 │
│ 6           │ VARCHAR     │ GSN                │ GSN         │             1 │ NULL               │ NULL               │ NULL                │ NULL               │ NULL                │ 129657 │           99.24 │
│ 7           │ VARCHAR     │ CRN                │ HCN         │             2 │ NULL               │ NULL               │ NULL                │ NULL               │ NULL                │ 129657 │           98.88 │
│ 8           │ DOUBLE      │ 1001.0             │ 99999.0     │          6687 │ 55235.1683404469   │ 26569.622608861097 │ 33935.722990246235  │ 58160.921869566904 │ 72584.6584736155    │ 129657 │           93.86 │
└─────────────┴─────────────┴────────────────────┴─────────────┴───────────────┴────────────────────┴────────────────────┴─────────────────────┴────────────────────┴─────────────────────┴────────┴─────────────────┘
```

**Reply by xevix** [reply]
@pdet Do you think it makes sense to not use the `BaseScanner` for this, and instead make a new scanner? Things like quoted strings don't need to exist in fixed-width since a string is bounded by its width already. Working on figuring out where to begin tackling this one.

**Comment by rustyconover** [comment]
I have some work on this already started.  Stay tuned.

**Reply by xevix** [reply]
Gotcha, I’ll pause on this then.

**Reply by xevix** [reply]
Hey @rustyconover , are you still working on this?

**Comment by rustyconover** [comment]
Hi All, 

I've released https://github.com/Query-farm/vgi-fixedformat:

```sql
INSTALL vgi FROM community;
ATTACH 'fixed' (TYPE vgi, LOCATION 'github-auto://Query-farm/vgi-fixedformat@v0.8.0');
SELECT unpack_fixed('JOHN      00042', 'name:A10 qty:9(5)');  -- {'name': JOHN, 'qty': 42}
```

It can read and write fixed formatted files very easily.

Rusty
